arangod/Transaction/Methods.cpp

```diff
@@ -1415,229 +1620,198 @@ Future<OperationResult> transaction::Methods::modifyCoordinator(
-        // expect a following term id or send a refusal
-        if (!options.isRestore) {
-          sendRefusal |= (options.isSynchronousReplicationFrom.find('_') ==
-                          std::string::npos);
-        }
-      }
-      if (sendRefusal) {
-        return OperationResult(
-            ::buildRefusalResult(
-                *collection,
-                (operation == TRI_VOC_DOCUMENT_OPERATION_REPLACE ? "replace"
-                                                                 : "update"),
-                options, theLeader),
-            options);
-      }
+  std::shared_ptr<std::vector<ServerID> const> followers;
 
-      // we are a valid follower. we do not need to send a proper result with
-      // _key, _id, _rev back to the leader, because it will ignore all these
-      // data anyway. it is sufficient to send headers and the proper error
-      // codes back.
-      options.silent = true;
-    }
-  }  // isDBServer - early block
+  Result res = determineReplicationTypeAndFollowers(
+      *collection, isUpdate ? "update" : "replace", newValue, options,
+      replicationType, followers);
 
-  TRI_ASSERT((replicationType == ReplicationType::LEADER) ==
-             (followers != nullptr));
-  TRI_ASSERT(!options.silent || replicationType != ReplicationType::LEADER ||
-             followers->empty());
-
-  // Update/replace are a read and a write, let's get the write lock already
-  // for the read operation:
-  //  Result lockResult = lockRecursive(cid, AccessMode::Type::WRITE);
-  //
-  //  if (!lockResult.ok() && !lockResult.is(TRI_ERROR_LOCKED)) {
-  //    return OperationResult(lockResult);
-  //  }
-
-  VPackBuilder resultBuilder;  // building the complete result
-  ManagedDocumentResult previous;
-  ManagedDocumentResult result;
-
-  if (options.validate && !options.isRestore &&
-      options.isSynchronousReplicationFrom.empty()) {
-    options.schema = collection->schema();
-  } else {
-    options.schema = nullptr;
+  if (res.fail()) {
+    return futures::makeFuture(OperationResult(std::move(res), options));
   }
 
-  [[maybe_unused]] size_t numExclusions = 0;
+  // set up batch options
+  BatchOptions batchOptions;
+  if (!options.isRestore && options.isSynchronousReplicationFrom.empty()) {
+    if (_state->isDBServer()) {
+      batchOptions.validateShardKeysOnUpdateReplace =
+          (collection->shardKeys().size() > 1 ||
+           collection->shardKeys()[0] != StaticStrings::KeyString);
+      batchOptions.validateSmartJoinAttribute =
+          collection->hasSmartJoinAttribute();
+    }
+    if (options.validate) {
+      batchOptions.schema = collection->schema();
+    }
+    std::shared_ptr<ComputedValues> cv = collection->computedValues();
+    if (cv != nullptr && ((isUpdate && cv->mustComputeValuesOnUpdate()) ||
+                          (!isUpdate && cv->mustComputeValuesOnReplace()))) {
+      batchOptions.computedValues = std::move(cv);
+    }
+  }
```

Comment:
`insertLocal` has a very similar code block. Can we pull out a helper function to reduce code duplication?